### PR TITLE
refactor(http): tree-shakable error on JSONP request

### PR DESCRIPTION
**Pull request by JeanMeche**
This commit provides a tree shakable error message (+doc) when a JSONP request is made without loading the `HttpClientJsonpModule`.

Expect a followup PR(s) for the other errors in that package. 

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

### CODE REVIEW

File: aio/content/errors/NG2800.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+@name Missing HttpClientJsonpModule
```
Comment (AndrewKushnir) [inline aio/content/errors/NG2800.md:1]:
```suggestion
@name Missing JSONP support in HttpClient configuration
```

File: aio/content/errors/NG2800.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+@category runtime
+@shortDescription HttpClientJsonpModule is required for JSONP support
```
Comment (AndrewKushnir) [inline aio/content/errors/NG2800.md:3]:
```suggestion
@shortDescription Missing JSONP support in HttpClient configuration
```

File: aio/content/errors/NG2800.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+@description
+You are attempting to make a `JSONP` request without providing the `HttpClientJsonpModule`.
```
Comment (AndrewKushnir) [inline aio/content/errors/NG2800.md:6]:
```suggestion
Angular produces this error when you try to make a `JSONP` request without providing the necessary support for it in the `HttpClient` configuration.

To enable `JSONP` support, you can:
- either add the `withJsonpSupport()` as an argument during the `provideHttpClient` function call (e.g. `provideHttpClient(withJsonpSupport())`) when `bootstrapApplication` is used
- or import the `HttpClientJsonpModule` in your root AppModule, when NgModule-based bootstrap is used.
```

File: aio/content/errors/NG2800.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+<code-example format="typescript" language="typescript">
+import {provideHttpClient, withJsonpSupport} from  '@angular/common/http';
+
+&commat;Component({
+  providers: [
+    provideHttpClient(withJsonpSupport()),
+  ]
+})
+</code-example>
```
Comment (AndrewKushnir) [inline aio/content/errors/NG2800.md:25]:
```suggestion
Make sure that the JSONP support is added into your application either by calling the `withJsonpSupport()` function (when the `provideHttpClient()` is used) or importing the `HttpClientJsonpModule` module as described above.
```

File: packages/common/http/src/xhr.ts [hunk 5]
```diff
@@ -52,8 +53,10 @@ export class HttpXhrBackend implements HttpBackend {
     // Quick check to give a better error message when a user attempts to use
     // HttpClient.jsonp() without installing the HttpClientJsonpModule
     if (req.method === 'JSONP') {
-      throw new Error(
-          `Attempted to construct Jsonp request without HttpClientJsonpModule installed.`);
+      throw new RuntimeError(
+          RuntimeErrorCode.MISSING_JSONP_MODULE,
+          (typeof ngDevMode === 'undefined' || ngDevMode) &&
+              `Attempted to construct Jsonp request without the HttpClientJsonpModule provided. Provide this module with \`provideHttpClient(withJsonpSupport())\``);
```
Comment (AndrewKushnir) [inline packages/common/http/src/xhr.ts:59]:
```suggestion
              `Attempted to construct Jsonp request without enabling support for it. To fix the problem, either add the \`withJsonpSupport()\` call (if `provideHttpClient()` is used) or import the \`HttpClientJsonpModule\` in the root NgModule.`);
```

**Review by AndrewKushnir — COMMENTED** [review]
@JeanMeche looks great, thanks for improving the code and the docs! 👍 

I've left some comments, please let me know if you have any questions.

Thank you.

File: packages/common/http/src/xhr.ts (hunk 5 above)
Comment (AndrewKushnir) [reply packages/common/http/src/xhr.ts:59]: @JeanMeche it looks like the error message remains the same, just wanted to check if you pushed the most recent version of this file to the branch?

**Comment by JeanMeche** [comment]
Is it me or aio tests breakage are unrelated to the changes here ?

**Comment by AndrewKushnir** [comment]
@JeanMeche could you please rebase this PR when you get a chance (there is a conflict with the main branch at this moment)?

**Comment by AndrewKushnir** [comment]
[Presubmit](https://fusion2.corp.google.com/presubmit/528690008/OCL:528690008:BASE:528983315:1683089793434:8b0484b8/targets) is "green".

**Comment by AndrewKushnir** [comment]
@JeanMeche the `test_aio` and `test_aio_local` CI jobs are failing, but it's a bit surprising, since the changes in this PR should not affect any AIO behavior. I'm wondering if the rebase just happened at the time those CI jobs were failing in main. Could you please try to rebase on top of the most recent main branch once again (that'd kick off CI as well)? Thank you.

**Comment by JeanMeche** [comment]
@AndrewKushnir I gave it 2 tries with no luck. I'm a bit clueless about what is responsible here.

File: aio/content/errors/NG2800.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
+@name JSONP support in HttpClient configuration
```
Comment (AndrewKushnir) [inline aio/content/errors/NG2800.md:1]: I think the file should be named as `NG02800.md` (an extra `0` after `NG`), all runtime codes have leading `0` (i.e. `NG0` as a prefix).

File: packages/common/http/src/errors.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+/**
+ * The list of error codes used in runtime code of the `common/http` package.
+ * Reserved error code range: 2800-2899.
+ */
+export const enum RuntimeErrorCode {
+
```
Comment (AndrewKushnir) [inline packages/common/http/src/errors.ts:14]: Let's remove this extra line.
```suggestion
```

**Review by AndrewKushnir — COMMENTED** [review]
@JeanMeche I've left a couple comments, but they are unlikely solve the problem with CI.

My proposal would be to try simplifying the code as much as possible for testing purposes (may be in a new draft PR?), i.e. make the `.md` file as simple as possible, replace the error message in the `throw` with a simple string, etc.

**Comment by AndrewKushnir** [comment]
@JeanMeche you could also try re-creating the PR (create a new one), there is a chance that things just got cached on CI at a "broken" state.

**Comment by JeanMeche** [comment]
A brand new PR did the job, #50376

**Comment by AndrewKushnir** [comment]
> A brand new PR did the job, https://github.com/angular/angular/pull/50376

Yay! 🎉

**Pull request closed without merging**